dvc/repo/data.py

```diff
@@ -10,124 +9,98 @@
-    from dvc_data.hashfile.hash_info import HashInfo
-    from dvc_data.hashfile.obj import HashFile
+    from dvc_data.index import DataIndex
 
 
 def posixpath_to_os_path(path: str) -> str:
     return path.replace(posixpath.sep, os.path.sep)
 
 
-def _in_cache(obj: "HashInfo", cache: "HashFileDB") -> bool:
-    from dvc_objects.errors import ObjectFormatError
+def _adapt_typ(typ):
+    from dvc_data.index.diff import ADD, DELETE, MODIFY
 
-    assert obj.value
-    try:
-        cache.check(obj.value)
-        return True
-    except (FileNotFoundError, ObjectFormatError):
-        return False
+    if typ == MODIFY:
+        return "modified"
 
+    if typ == ADD:
+        return "added"
 
-def _shallow_diff(
-    root: str,
-    old: Optional["HashInfo"],
-    new: Optional["HashInfo"],
-    cache: "HashFileDB",
-) -> Dict[str, List[str]]:
-    d = {}
-
-    old_root = new_root = root
-    if old and old.isdir:
-        old_root = os.path.sep.join([root, ""])
-    if new and new.isdir:
-        new_root = os.path.sep.join([root, ""])
-
-    if old and not _in_cache(old, cache):
-        d["not_in_cache"] = [old_root]
-
-    if not old and not new:
-        return d
-    if not new:
-        return {"deleted": [old_root], **d}
-    if not old:
-        return {"added": [new_root], **d}
-    if old != new:
-        return {"modified": [new_root], **d}
-    return {"unchanged": [new_root], **d}
-
-
-def _granular_diff(
-    root: str,
-    old_obj: Optional["HashFile"],
-    new_obj: Optional["HashFile"],
-    cache: "HashFileDB",
-) -> Dict[str, List[str]]:
-    from dvc_data.hashfile.diff import ROOT
-    from dvc_data.hashfile.diff import diff as odiff
+    if typ == DELETE:
+        return "deleted"
 
-    def path_join(root: str, *paths: str, isdir: bool = False) -> str:
-        if not isdir and paths == ROOT:
-            return root
-        return os.path.sep.join([root, *paths])
+    return typ
 
-    diff_data = odiff(old_obj, new_obj, cache)
 
-    output: Dict[str, List[str]] = defaultdict(list)
-    for state in ("added", "deleted", "modified", "unchanged"):
-        items = getattr(diff_data, state)
-        for item in items:  # pylint: disable=not-an-iterable
-            entry = item.old if state == "deleted" else item.new
-            isdir = entry.oid.isdir if entry.oid else False
+def _adapt_path(change):
+    isdir = False
+    if change.new and change.new.meta:
+        isdir = change.new.meta.isdir
+    elif change.old and change.old.meta:
+        isdir = change.old.meta.isdir
+    key = change.key
+    if isdir:
+        key = (*key, "")
+    return os.path.sep.join(key)
 
-            path = path_join(root, *entry.key, isdir=isdir)
-            output[state].append(path)
-            if not item.old.in_cache and state != "added":
-                output["not_in_cache"].append(path)
-    return dict(output)
 
+def _diff(
+    old: "DataIndex",
+    new: "DataIndex",
+    *,
+    granular: Optional[bool] = False,
```

Comment:
```suggestion
    granular: bool = False,
```

Reply:
using None is also absolutely valid.

Reply:
It's as valid as passing an empty string or an empty list. It's better to keep types simple.

Reply:
Optional boolean is almost always a wrong choice. It could be rarely useful for ternary logic, or to denote the absence of values. But for truthy/non-truthy cases, it's always better to use `bool` to communicate the intention.